### ISSUE TITLE
Modelling many to many, and one to one relationships

### ISSUE THREAD

**Issue by wstrange** [stablekernel/aqueduct#149]
Not sure if this a documentation issue, or a feature request :-)

How does one model the following:

1) Many to Many.  For example,  Golfers belong to many golf clubs, and golf clubs have many members  (many to many from clubs to golfers) 

2) one to one - with more than mapping. Example:  golfers have an address, and golf clubs have an address.  I want a single address object in my mappings. Not  golfer_address, club_address.
Hibernate has "embedded" objects - where (for example), address is embedded as additional columns in the SQL. That would also be nice...

**Comment by joeconwaystk** [comment]
1. You can create a join table by subclassing ManagedObject and having two to-many relationships to the related entities. Currently, you have to fetch rows from the join table in one query and then the other side of the relationship in another. We'll improve this at some point, and I don't think it'll be particularly challenging. This will get added to the 1.0 roadmap.

2. Not yet, and hadn't thought of it, but it sounds like a good feature. The underlying database driver would need to support some sort of column object store (PostgresSQL would likely be JSONB, for example.) I'll add this to the 2.0 or later roadmap.

**Comment by ronlobo** [comment]
Little example for 2., which is called polymorphic models for a similar framework to aqueduct, Laravel (https://laravel.com/docs/5.3/eloquent-relationships#polymorphic-relations)

In this case, we would have:

addresses
 - id
 - ...
 - addressable_id (foreign key)
 - addressable_type (table name)

Thoughts?

**Comment by wstrange** [comment]
JPA seems to have figured this out, so I'd consider taking a look at it:
https://en.wikibooks.org/wiki/Java_Persistence/OneToOne 

https://en.wikibooks.org/wiki/Java_Persistence/ManyToMany

https://en.wikibooks.org/wiki/Java_Persistence/Embeddables

**Comment by joeconwaystk** [comment]
@ronlobo So that's a neat feature. Coincidentally, I was working on a feature that will serve as the foundation to this (PR #153) and so maybe it makes sense to extend that behavior one step further.

This PR lets a relationship property refer to an interface, which can then be implemented by a concrete persistent type in some other package. (So you can have objects that maintain referential integrity, but are declared in different packages.) The data model builder checks to see if there is exactly one concrete implementation of the persistent type. But, to support this feature, the builder can allow for multiple implementations and add a column to keep track of the type. 

So to the mirror the example in Laravel, this might look like:

```dart
class Comment extends ManagedObject<_Comment> implements _Comment {}
class _Comment {
  @managedPrimaryKey 
  int id;

  String body;
  
  @ManagedDeferredRelationship(/*deleteRule, required*/)
  Commentable  owner;
}

// We'd have to enforce the primary key to be the same type 
// for all commentables, but don't want to enforce the name.
class Commentable {
 // We don't really need anything in here..
}

class Video extends ... {}
class _Video extends Commentable {
...
}

class Post extends ... {}
class _Post extends Commentable {
...
}
```

**Comment by joeconwaystk** [comment]
Also, question for @ronlobo on that example in Laravel... is there no referential integrity between the the entities? If commentable_id can reference a column from either posts or videos, how is the underlying database managing delete rules?

**Comment by ronlobo** [comment]
@joeconwaystk Usually, you would define cascading within the Laravel migrations which will set it on the database level when running the migrations. But polymorphic relationships are a feature of the ORM Eloquent which is used in Laravel, which means there is no referential integrity between polymorphic entities. For example when deleting a post, you would have to override the Eloquent delete method in your post model, get the attached comments and delete those on its own. I am not quite sure if there is anything planned to make handling this easier on the Eloquent side of things.

**Comment by joeconwaystk** [comment]
This is addressed by `Document` properties in 3.0
